doc/barclamp.md

```diff
@@ -0,0 +1,1321 @@
+
+```
+
+The corresponding file from the Barbican barclamp is
+[here](https://github.com/crowbar/crowbar-openstack/blob/8bebf8a379ebea8ef462ad49746dda6d36a3c46d/crowbar_framework/config/locales/barbican/en.yml).
+It contains field descriptions for a bunch of fields already, since this will
+make it easier to add UI elements later. Feel free to prepare your own barclamp
+in a similar manner.
+
+##### Minimal Model Class
+
+Location: `crowbar_framework/app/models/mybarclamp_service.rb`
+
+With the view in place we need a model class to pre-process the proposal
+information and enforce constraints that cannot be expressed using the
+declarative means available in `template-mybarclamp.schema`, i.e. constraints
+that need program logic to check. Just like in the data bag sections we will
+again use the Barbican barclamp's proposal as an example and modify it bit by
+bit, so as a first step, copy `crowbar_framework/app/models/barbican_service.rb`
+to `crowbar_framework/app/models/mybarclamp_service.rb` and globally replace
+`barbican` by `mybarclamp` and `Barbican` by `Mybarclamp`, respectively.
+
+This will leave us with with something like this in the first few lines:
+
+```
+class MybarclampService < PacemakerServiceObject
+  def initialize(thelogger)
+    @bc_name = "mybarclamp"
+    @logger = thelogger
+  end
+```
+
+The corresponding lines from the Barbican barclamp's controller begin
+[here](https://github.com/crowbar/crowbar-openstack/blob/8bebf8a379ebea8ef462ad49746dda6d36a3c46d/crowbar_framework/app/models/barbican_service.rb#L17) and end
+[here](https://github.com/crowbar/crowbar-openstack/blob/8bebf8a379ebea8ef462ad49746dda6d36a3c46d/crowbar_framework/app/models/barbican_service.rb#L21).
+
+The next thing we need to change is the `role_constraints` method. This method
+tells Crowbar on how many nodes the service can be deployed, what types of node
+are valid and on which platform it may run. All the roles you defined in the
+chef sections need to be accounted for by this method. This means that we will
+need to add an entry for the `mybarclamp-compute` role. Below you will see a
+commented example:
+
+```
+    def role_constraints
+      {
+        "mybarclamp-controller" => {
+          "unique" => false,   # Set to `true` if this can only be deployed to one node.
+          "count" => 3,        # Number of allowed instances ('-1' means infinite)
+          "cluster" => true,   # Whether this role can be clustered (important for HA later on).
+          "admin" => false,    # Whether this role can be assigned to the Crowbar admin node.
+          "exclude_platform" => {
+            "suse" => "< 12.1",  # Do not deploy on SUSE version 12.1 and lower.
+            "windows" => "/.*/"  # Do not deploy on any Windows platforms.
+          }
+        },
+         "mybarclamp-compute" => {
+          "unique" => false,
+          "count" => -1,   # Can be deployed to an arbitrary number of Compute nodes
```

Comment:
Yes...I wanted to have a concrete usage example as well, in case somebody missed the comment above.